python/cudf/cudf/core/buffer/spill_manager.py

```diff
@@ -0,0 +1,363 @@
+
+        Return
+        ------
+        int
+            Number of bytes spilled.
+        """
+        for buf in self.base_buffers(order_by_access_time=True):
+            if buf.lock.acquire(blocking=False):
+                try:
+                    if not buf.is_spilled and buf.spillable:
+                        buf.__spill__(target="cpu")
+                        return buf.size
+                finally:
+                    buf.lock.release()
+        return 0
+
+    def spill_to_device_limit(self, device_limit: int = None) -> int:
+        limit = (
+            self._device_memory_limit if device_limit is None else device_limit
+        )
+        if limit is None:
+            return 0
+        ret = 0
+        while True:
+            unspilled = sum(
+                buf.size for buf in self.base_buffers() if not buf.is_spilled
+            )
+            if unspilled < limit:
+                break
+            nbytes = self.spill_device_memory()
+            if nbytes == 0:
+                break  # No more to spill
+            ret += nbytes
+        return ret
+
+    def lookup_address_range(
+        self, ptr: int, size: int
+    ) -> List[SpillableBuffer]:
+        ret = []
+        for buf in self.base_buffers():
+            if buf.is_overlapping(ptr, size):
+                ret.append(buf)
+        return ret
+
+    def __repr__(self) -> str:
+        spilled = sum(
+            buf.size for buf in self.base_buffers() if buf.is_spilled
+        )
+        unspilled = sum(
+            buf.size for buf in self.base_buffers() if not buf.is_spilled
+        )
+        unspillable = 0
+        for buf in self.base_buffers():
+            if not (buf.is_spilled or buf.spillable):
+                unspillable += buf.size
+        unspillable_ratio = unspillable / unspilled if unspilled else 0
+
+        return (
+            f"<SpillManager spill_on_demand={self._spill_on_demand} "
+            f"device_memory_limit={self._device_memory_limit} | "
+            f"{format_bytes(spilled)} spilled | "
+            f"{format_bytes(unspilled)} ({unspillable_ratio:.0%}) "
+            f"unspilled (unspillable)>"
+        )
+
+
+# TODO: do we have a common "get-value-from-env" in cuDF?
+def _env_get_int(name, default):
+    try:
+        return int(os.getenv(name, default))
+    except (ValueError, TypeError):
+        return default
+
+
+def _env_get_bool(name, default):
+    env = os.getenv(name)
+    if env is None:
+        return default
+    as_a_int = _env_get_int(name, None)
+    env = env.lower().strip()
+    if env == "true" or env == "on" or as_a_int:
+        return True
+    if env == "false" or env == "off" or as_a_int == 0:
+        return False
+    return default
+
+
+def _get_manager_from_env() -> Optional[SpillManager]:
+    if not _env_get_bool("CUDF_SPILL", False):
+        return None
+    return SpillManager(
+        spill_on_demand=_env_get_bool("CUDF_SPILL_ON_DEMAND", True),
```

Comment:
Is there ever a reason to disable spilling on demand aside from debugging?

Reply:
I guess there are performance considerations here. What is the alternative to on-demand spilling? If it's off, when do we spill?

Reply:
> If it's off, when do we spill?

Then no spilling by default but setting `CUDF_SPILL_DEVICE_LIMIT` will trigger spilling.